### PR TITLE
core: optimize the processing speed of RegionHeartbeat

### PR DESCRIPTION
**Pull request by HunDunDM**
### What problem does this PR solve?

* The current heartbeat processing speed can be increased.

### What is changed and how it works?

* Move `hotStat` out of the read lock
* Move `storage.DeleteRegion` out of the write lock
* Simple processing logic is used for RegionHeartbeat that only has statistics updated.

### Check List

Tests

- Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: server/core/region.go [hunk 1]
```diff
@@ -518,98 +536,114 @@ func (r *RegionsInfo) TreeLen() int {
...
-				r.tree.updateStat(regionOld.region, region)
-				regionOld.region = region
-				treeNeedAdd = false
-			}
-		}
+	var origin *RegionInfo
+	if item == nil {
+		item = r.regions.AddNew(region)
+	} else {
+		origin = item.region
```
Comment (nolouch) [inline server/core/region.go:577]: add a comment about update region in map.

File: server/core/region.go [hunk 2]
```diff
@@ -488,24 +493,37 @@ func NewRegionsInfo() *RegionsInfo {
...
+	var item *regionItem
+	rangeChanged := true
+	peersChanged := true
+	if item = r.regions.Get(region.GetID()); item != nil {
+		origin := item.region
+		rangeChanged = !bytes.Equal(origin.GetStartKey(), region.GetStartKey()) ||
+			!bytes.Equal(origin.GetEndKey(), region.GetEndKey())
+		peersChanged = r.shouldRemoveFromSubTree(region, origin)
+		if rangeChanged {
+			r.tree.remove(origin)
```
Comment (nolouch) [inline server/core/region.go:513]: add a comment why no need to remove from region map.

**Review by nolouch — COMMENTED** [review]
rest LGTM

**Review by nolouch — APPROVED** [review]
Lgtm

**Comment by nolouch** [comment]
PTAL @rleungx @lhy1024

**Review by lhy1024 — APPROVED** [review]

**Pull request merged**